node/src/pow/blake3_pow.rs

```diff
@@ -139,3 +239,53 @@ fn hash_passes_threshold_test(hash: blake3::Hash, threshold: Threshold) -> bool
     let hash_value = Threshold::from_big_endian(hash.as_bytes());
     hash_value <= threshold
 }
+
+/// Calculates the difficulty for the next block based on the window of the previous blocks
+///
+/// `avg` - the average difficulty of the blocks in the window
+/// `time_observed` - the total time it took to create the blocks in the window
+/// `time_target` - the total time it should take to create the blocks in the window
+fn next_difficulty(avg: Difficulty, time_observed: u64, time_target: u64) -> Difficulty {
+    // This is safe, because difficulty has only its low 192 bits used
+    let new_raw = avg * time_target / time_observed;
+    if new_raw > avg {
+        let delta = new_raw - avg;
+        let damped_delta = delta / ADJUST_DIFFICULTY_DAMPING;
+        let new_damped = avg + damped_delta;
+        let new_max = avg * ADJUST_DIFFICULTY_CLAMPING;
+        new_damped.min(new_max)
+    } else {
+        let delta = avg - new_raw;
+        let damped_delta = delta / ADJUST_DIFFICULTY_DAMPING;
+        let new_damped = avg - damped_delta;
+        // Clamping matters only when ADJUST_DIFFICULTY_CLAMPING > ADJUST_DIFFICULTY_DAMPING
+        let new_min = avg / ADJUST_DIFFICULTY_CLAMPING;
+        new_damped.max(new_min)
+    }
+}
+
+#[cfg(test)]
+mod test {
+    use super::*;
+
+    #[test]
+    fn next_difficulty() {
+        assert_next_difficulty(67, 100, 100, 0);
+        assert_next_difficulty(90, 100, 100, 70);
+        assert_next_difficulty(100, 100, 100, 100);
+        assert_next_difficulty(110, 100, 100, 130);
+        assert_next_difficulty(199, 100, 100, 399);
+        assert_next_difficulty(200, 100, 100, 400);
+        assert_next_difficulty(200, 100, 100, 500);
+    }
+
+    fn assert_next_difficulty(expected: u64, avg: u64, time_observed: u64, time_target: u64) {
+        let expected = U256::from(expected);
+        let actual = super::next_difficulty(U256::from(avg), time_observed, time_target);
```

Comment:
`super` is not needed.
```suggestion
        let actual = next_difficulty(U256::from(avg), time_observed, time_target);
```

Reply:
It is, because otherwise we're calling the test function. We can either have `super::` or `_tests` suffix in a test name.